pkg/agent/plugin/workloadattestor/systemd/systemd_posix.go

```diff
@@ -68,11 +68,11 @@ func (p *Plugin) Attest(ctx context.Context, req *workloadattestorv1.AttestReque
 }
 
 func getSystemdUnitInfo(ctx context.Context, pid uint) (*DBusUnitInfo, error) {
+	// Do not close this connection because it is shared and will autoclose on errors
 	conn, err := dbus.SystemBus()
 	if err != nil {
 		return nil, status.Errorf(codes.Internal, "failed to open dbus connection: %v", err)
 	}
-	defer conn.Close()
```

Comment:
If there is no error, when would be closed the connection?
I see that the connection [is closed when the context is done](https://github.com/godbus/dbus/blob/v5.1.0/conn.go#L304), but the context used [would be a new context](https://github.com/godbus/dbus/blob/v5.1.0/conn.go#L284) and not the one from this function.

Reply:
Since it is a background context which is [never cancelled](https://pkg.go.dev/context#Background), it looks like the connection is never closed. Or did I get that wrong?

Reply:
Right, I don't think that the connection will be closed, which is a problem.
We will need to avoid using the shared dbus connection, since it doesn't seem that there is a clear way to properly handle the closure of connections when there are concurrent requests.
I think that the right solution here will involve maintaining a connection at the plugin level.

Reply:
Do you mean a shared connection at the plugin level that's never closed? It's slightly better because it'll protect the plugin from other code inadvertently closing the connection.

Reply:
Yes, I mean having a single connection at the plugin level that's closed when the plugin closes. The same connection can be used by the different attestation calls.

Reply:
Thank you, @amartinezfayo! I have changed this to use a plugin level shared connection. I could not find the lifecycle mechanism to close the connection on plugin close (probably because I am fairly new to spire code base and go). What am I missing?

Reply:
When plugins implement the io.Closer interface, the Close function is invoked before the plugin is unloaded. Since this is [documented](https://github.com/spiffe/spire-plugin-sdk/blob/main/docs/AUTHORING.md#cleanup) in the SPIRE Plugin SDK, it may be difficult to find it :)
You just need to add a function `func (p *Plugin) Close() error {...}` that closes the connection, and it will be automatically called.